`break with` and `continue with` for modifying results array in a loop

edemaine opened the pull request:
Fixes #1395

TODO: Documentation

An issue with TypeScript is clear in the `loop` example: we have `let results = []` but later `results = 'done'`. Perhaps we could detect specifically a `loop` that has no plain `break` and thus will only be exited via `break with`, and then we could just write `let results`.

STRd6 reviewed (approved):
Looks great!

One solution to help TypeScript is to declare the results as `let results;results=[]`. This gives it the proper union type of all possible exits from the loop.

vendethiel commented on test/for.civet:
```diff
@@ -774,10 +774,23 @@ describe "for", ->
+  testCase """
+    break/continue with outside expression context
+    ---
+    for x of y
```
Maybe this should warn?

edemaine replied:
Hmm, good question. It does feel a little worrying throwing away code (the `with` arguments), though I think it's nicer than failing, in particular as you're transforming code to/from expression mode.

We don't currently have a warning system. @STRd6 Should we implement one? Perhaps `type: "Warning"` nodes, or `warning: true`/`subtype: "Warning"` on `type: "Error"` nodes? This would require some restructuring of current error handling, to report when possible but fail only in certain cases.

STRd6 replied:
Seems like something we'd like eventually. We can use a subtype on Error nodes for now.

edemaine replied:
Sounds good. If it's OK with you, I'd like to leave this for a future project. Actually checking for `break with`/`continue with` that isn't consumed in a loop is trickier.

I also realize that `break :label with foo` doesn't currently work. (Currently you can only use `break with`/`continue with` to modify the parent loop.) Maybe also good for a future PR.

edemaine commented:
> One solution to help TypeScript is to declare the results as `let results;results=[]`. This gives it the proper union type of all possible exits from the loop.

Good idea! Implemented.

I also removed `results=[]` in the special case of `loop` containing `break with` but not `break`.
